### ISSUE TITLE
Narrowing of `int | Sequence[str]` with `isinstance(Sequence)` is not Sequence[str]

### ISSUE THREAD

**Issue by henriquegemignani** [astral-sh/ty#2493]
### Summary

Playground: https://play.ty.dev/6143bcb7-6aa9-405d-ba62-33a4ea7e764d

Code:
```python
from collections.abc import Sequence

type JsonType_RO = int | Sequence[str]


def encode_a(value: JsonType_RO) -> None:
    if isinstance(value, Sequence):
        sequence_value: Sequence[str] = value

def encode_b(value: JsonType_RO) -> None:
    if isinstance(value, Sequence[str]):
        sequence_value: Sequence[str] = value
```
Errors:
```
Object of type `(int & Sequence[object]) | Sequence[str]` is not assignable to `Sequence[str]` (invalid-assignment) [Ln 8, Col 41]
Object of type `JsonType_RO` is not assignable to `Sequence[str]` (invalid-assignment) [Ln 12, Col 41]
```

The main case is `encode_a`. I'd expect `value` to be narrowed to `Sequence[str]`, but that's not the case. As far as I could find, that should be valid typing. I tried to find some related issue, but no luck with search.

`encode_b` is just an attempt to investigate further, but it's interesting how that causes ty to not narrow at all.

### Version

ty 0.0.11 (830cb9cc6 2026-01-09)

**Comment by AlexWaygood** [comment]
Thanks! This is basically the same as #1578. ty is warning you here that it's entirely possible for an object to be an instance of both `int` and `Sequence` simultaneously: they're not disjoint types:

```py
>>> from typing import Sequence
>>> class Foo(int, Sequence): ...
... 
>>>
```

You can fix the diagnostic here by using `if not isinstance(value, int)` instead of `if isinstance(value, Sequence)`.

**Comment by henriquegemignani** [comment]
Is there a way to annotate that such subclass of `int` is not allowed? In the actual use in my code base it does already `if isinstance(value, int)`, but the actual type is `int | Mapping[str, JsonType_RO] | Sequence[JsonType_RO]` and adding a `not isinstance(value, Mapping)` doesn't work.

**Comment by AlexWaygood** [comment]
> Is there a way to annotate that such subclass of `int` is not allowed?

No, Python's type system doesn't currently support a way to annotate "exact types", and we don't have any extensions to the type system for that currently in ty. Such types would unfortunately not be very useful unless you used them _everywhere_, but in most cases it isn't desirable to distinguish between "any instance of `T`", and "any object that is an instance of _exactly_ `T`". So I think the use cases for types like this would actually turn out to be pretty limited, unfortunately.

If you're willing to use the `ty_extensions` module, you could do something like this:

```py
from ty_extensions import Intersection, Not
from collections.abc import Sequence

type JsonType_RO = Intersection[int, Not[Sequence[object]]] | Sequence[str]


def encode_a(value: JsonType_RO) -> None:
    if isinstance(value, Sequence):
        sequence_value: Sequence[str] = value
```

But note that the ty_extensions module doesn't actually exist at runtime right now; you need to import it in an `if TYPE_CHECKING` block if you want to use it in a `.py` file.

I'm curious to know more about this, though:

> adding a `not isinstance(value, Mapping)` doesn't work.

Why doesn't that work in this case for you? Can you give another minimal example? :-)

**Comment by henriquegemignani** [comment]
> Why doesn't that work in this case for you? Can you give another minimal example? :-)

Funnily enough in my actual code-base doing this worked, but I'm not sure why it doesn't in the minimal example. https://play.ty.dev/a28ea65f-d286-4653-9f32-ea2c5398488c

Edit: because I forgot an `else`...

**Comment by carljm** [comment]
I think the working version can even be simplified a bit more:

```py
from collections.abc import Sequence, Mapping

type JsonObject_RO = Mapping[str, "JsonType_RO"]
type JsonType_RO = int | JsonObject_RO | Sequence["JsonType_RO"]


def encode_json_value(value: JsonType_RO) -> None:
    if isinstance(value, int):
        int_value: int = value

    elif not isinstance(value, Mapping):
        sequence_value: Sequence[JsonType_RO] = value

    elif not isinstance(value, Sequence):
        mapping_value: JsonObject_RO = value

    else:
        raise RuntimeError()
```

https://play.ty.dev/8bb7aa29-e0f6-43b0-aefc-1958933ec6b5
